packages/ensnode-sdk/src/registrar-actions/zod-schemas.ts

```
@@ -0,0 +1,127 @@
+import { decodeEncodedReferrer, ENCODED_REFERRER_BYTE_LENGTH } from "@namehash/ens-referrals";
+import type { Address } from "viem";
+import z from "zod/v4";
+import type { ParsePayload } from "zod/v4/core";
+
+import {
+  makeDurationSchema,
+  makeEventRefSchema,
+  makeHexStringSchema,
+  makeLowercaseAddressSchema,
+  makePriceEthSchema,
+} from "../internal";
+import { CurrencyIds } from "../shared";
+import { type RegistrarAction, RegistrarActionTypes, RegistrarEventNames } from "./types";
+
+/** Invariant: total is sum of baseCost and premium */
+function invariant_registrarActionTotalIsSumOfBaseCostAndPremium(
+  ctx: ParsePayload<Pick<RegistrarAction, "baseCost" | "premium" | "total">>,
+) {
+  const registrarAction = ctx.value;
+
+  if (
+    registrarAction.total.amount !==
+    registrarAction.baseCost.amount + registrarAction.premium.amount
```

We need to define a utility function for something like `addPrices` that takes as input two `Price` values and returns back a single `Price` value.

This function should throw an error if the currency of the provided prices are not equal.

You can move this into a separate follow-up issue if you'd like.